File: MathAndPrimeNumbers/PerfectNumber/perfect.py
from math import sqrt

class Solution1:
    def checkPerfectNumber(self, num: int) -> bool:
        factorSum = 1
        for i in range(2,int(sqrt(num))+1):
            if num%i == 0 and i!=(num//i):
                factorSum += (i + num//i)
        if int(sqrt(num))*int(sqrt(num)) == num:
            factorSum += int(sqrt(num))
        if factorSum == num: return True
        return False
    

class Solution2:
    def checkPerfectNumber(self, num: int) -> bool:
        if num==1: return False
        factorSum = 0
        for i in range(1,int(sqrt(num)+1)):
            if num % i == 0: factorSum += i
        for i in range(int(sqrt(num)),1,-1):
            if num % (num//i) == 0: factorSum += (num//i)
        if factorSum == num: return True
        return False

File: MathAndPrimeNumbers/PerfectNumber/perfect.cpp
#include <cmath>
using namespace std;

class Solution1 {
public:
    bool checkPerfectNumber(int num) {
        int sum = 1;
        for(int i=2;i<sqrt(num);i++){
            if(num%i==0 and num/i != i)sum=(sum+i+num/i);
        }
        if(int(sqrt(num))*int(sqrt(num)==num))sum+=sqrt(num);
        return (sum==num);
    }
};

class Solution2 {
public:
    bool checkPerfectNumber(int num) {
        if (num==1) return false;
        int factorSum = 0;
        for(int i=1;i<sqrt(num);i++)
            if(num % i == 0) factorSum += i;
        for(int i=sqrt(num);i>1;i--)
            if(num % (num/i) == 0) factorSum += (num/i);
        if (factorSum == num) return true;
        return false;
    }
};
